contracts/ROSCA.sol

```diff
@@ -0,0 +1,222 @@
+*/
+contract ROSCA {
+  address constant WETRUST = 0x0;
+  uint constant MIN_CONTRIBUTION_SIZE = 1000000000000;  // 1e12
+  uint constant MAX_FEE_IN_THOUSANDTHS = 200;
+  address constant WETRUST_FEE_ADDRESS = 0x0;
+  uint constant MINIMUM_TIME_BEFORE_ROSCA_START = 1 days;
+  uint constant MINIMUM_PARTICIPANTS = 2;
+  uint16 constant EPOCH_TO_RUN = 1;
+  uint constant MIN_ROUND_PERIOD_IN_DAYS = 1;
+  uint constant MAX_ROUND_PERIOD_IN_DAYS = 30;
+  uint constant MIN_DISTRIBUTION_RATIO = 65;
+
+  event LogParticipantApplied(address user);
+  event LogParticipantApproved(address user);
+  event LogContributionMade(address user, uint amount);
+  event LogNewLowestBid(uint bid,address winnerAddress);
+  event LogRoundFundsReleased(address winnerAddress, uint amountInWei);
+  event LogFundsWithdrawal(address user, uint amount,address destination);
+  event LogStartOfPeriod(uint currentRound);
+
+  // state variables
+  uint16 currentEpoch;
+  uint16 numEpochs;     // number of Epochs this contract will run
+  uint16 roundPeriodInDays;
+  uint16 serviceFeeInThousandths;
+  uint16 currentRound;  // currentRound will be set to 0 when ROSCA is created and will turn to one when the ROSCA actually starts
+  uint8 minParticipants;
+  address foreman;    // address of the foreman
+  uint contributionSize;
+  uint startTime;     // startTime must be in timestamp
+
+  struct User{
+    uint contributed; // Total amount contributed
+    uint paid; // number of times bid Won
+    uint pendingWithdrawl; // how much they are allowed to withdraw, i.e if someone wins , their pendingWithdrawal will go up by the bid.
+    bool alive; // needed to check if a member is indeed a member
+  }
+  mapping(address => User) members; // using struct User to keep track of contributions and paid, allowance and etc.
+  address[] membersAddresses;    // this is the only way to iterate through all the member's address
+
+  mapping(address => bool) pendingJoinRequest; // this way , address can be used as index, if we use address[] , we'll have to go thru a whole array
+
+  // bidding related state variable
+  uint lowestBid;
+  address winnerAddress;
+
+
+  modifier onlyForeman {
+  if(msg.sender != foreman) throw;
+  _;
+  }
+  modifier onlyBeforeStart {
+  if(currentRound != 0) throw;
+  _;
+  }
+
+  /**
+    * Creates a new ROSCA and initializes the necessary variables, ROSCA doesnt start until the specified startTime
+    * Creator of the contract becomes foreman and also added as the first member of the ROSCA
+    */
+  function ROSCA(
+    uint16 roundPeriodInDays_,
+    uint contributionSize_,
+    uint8 minParticipants_,
+    uint startTime_,
+    uint16 feeInThousandths_)
+  {
+    if (roundPeriodInDays < MIN_ROUND_PERIOD_IN_DAYS || roundPeriodInDays > MAX_ROUND_PERIOD_IN_DAYS) throw;
+    roundPeriodInDays = roundPeriodInDays_;
+    if (contributionSize_ < MIN_CONTRIBUTION_SIZE) throw;
+    contributionSize = contributionSize_;
+
+    numEpochs = EPOCH_TO_RUN;
```

Comment:
i used the constant EPOCH_TO_RUN as check and took out any reference to numEpochs or currentEpoch